File: packages/ios/Sources/CrumbCore/CrumbJavaScriptCrashEnvelope.swift
#if canImport(Darwin)
import Darwin
#endif
import Foundation

package enum CrumbJavaScriptCrashEnvelopeBuilder {
    package static func build(
        settings: CrumbReportSettings,
        record: CrumbJavaScriptCrashRecord,
        runtime: CrumbReportRuntime
    ) throws -> CrumbSerializedReportEnvelope {
        let crash = record.crash
        let description = truncateUTF8(
            "\(crash.errorType): \(crash.message)",
            maximumBytes: 4_000
        )
        let customContext = record.customContext.filter {
            settings.customContext[$0.key] == $0.value
        }
        let envelope = JavaScriptCrashEnvelopeDTO(
            schemaVersion: "1.0",
            reportID: record.reportID,
            trigger: "javascript_crash",
            triggeredAt: crash.occurredAt,
            submittedAt: crash.occurredAt,
            release: ReleaseDTO(
                platform: "react_native",
                appVersion: record.appVersion,
                nativeBuild: record.nativeBuild,
                jsBundleVersion: record.bundleVersion,
                environment: record.environment
            ),
            application: record.applicationName.map { ApplicationDTO(name: $0) },
            sdk: SDKDTO(
                name: "crumb-react-native",
                version: CrumbSDKVersion.current,
                integration: "react_native"
            ),
            runtime: RuntimeDTO(
                os: "ios",
                osVersion: runtime.osVersion,
                deviceFamily: runtime.deviceFamily,
                locale: runtime.locale,
                timezone: runtime.timezone
            ),
            userInput: UserInputDTO(
                category: "javascript_crash",
                description: description
            ),
            javascriptCrash: JavaScriptCrashDTO(
                source: crash.source.rawValue,
                kind: crash.kind.rawValue,
                errorType: crash.errorType,
                message: crash.message,
                rawStack: crash.rawStack,
                deduplicationKey: crash.fingerprint ?? "js_unknown",
                nativeTerminationWrapper: crash.nativeTerminationWrapper,
                breadcrumbs: crash.breadcrumbs.map {
                    BreadcrumbDTO(
                        timestamp: $0.timestamp,
                        category: $0.category,
                        message: $0.message
                    )
                }
            ),
            diagnostics: DiagnosticsDTO(
                capturedAt: crash.occurredAt,
                location: "javascript_crash_recovery",
                process: ProcessDTO(
                    name: ProcessInfo.processInfo.processName,
                    id: processID()
                ),
                thermalState: "unavailable",
                threads: ThreadsDTO(count: 0, busiest: []),
                gpu: GPUDTO(status: "unavailable_on_demand"),
                network: NetworkDTO(
                    status: "unknown",
                    transport: "unknown",
                    cellularGeneration: nil,
                    isExpensive: false,
                    isConstrained: false,
                    healthCheck: nil
                ),
                logs: LogsDTO(
                    status: "unavailable",
                    sources: [],
                    entries: [],
                    truncated: false,
                    droppedEntryCount: 0,
                    failures: ["crash_recovery"]
                ),
                stackTraces: StackTracesDTO(
                    status: "unavailable",
                    scope: "none",
                    threads: [],
                    truncated: false,
                    unavailableReason: "crash_recovery"
                )
            ),
            privacy: PrivacyDTO(
                screenshotCapture: "not_applicable",
                screenshotMasking: "not_applicable",
                diagnosticsCapture: "crash_recovery",
                logCapture: "not_applicable",
                policyStatus: settings.policyStatus.rawValue,
                workspacePolicyVersion: settings.workspacePolicyVersion
            ),
            customContext: customContext.isEmpty ? nil : customContext,
            artifacts: []
        )

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.keyEncodingStrategy = .convertToSnakeCase
        encoder.outputFormatting = [.sortedKeys]
        return CrumbSerializedReportEnvelope(
            reportID: record.reportID,
            submittedAt: crash.occurredAt,
            data: try encoder.encode(envelope)
        )
    }

    private static func processID() -> Int {
        #if canImport(Darwin)
        return Int(getpid())
        #else
        return 1
        #endif
    }

    private static func truncateUTF8(_ value: String, maximumBytes: Int) -> String {
        guard value.utf8.count > maximumBytes else { return value }
        var result = ""
        for character in value {
            let candidate = result + String(character)
            if candidate.utf8.count + 3 > maximumBytes { break }
            result = candidate
        }
        return result + "…"
    }
}

package enum CrumbJavaScriptCrashRecovery {
    package static func recoverPending() async -> Int {
        guard let settings = try? Crumb.reportSettings() else { return 0 }
        return await recoverPending(
            store: .shared,
            queue: .shared,
            settings: settings,
            runtime: CrumbReportRuntime(
                osVersion: ProcessInfo.processInfo.operatingSystemVersionString,
                deviceFamily: "iOS",
                locale: Locale.current.identifier,
                timezone: TimeZone.current.identifier
            )
        )
    }

    package static func recoverPending(
        store: CrumbJavaScriptCrashStore,
        queue: CrumbReportQueue,
        settings: CrumbReportSettings,
        runtime: CrumbReportRuntime
    ) async -> Int {
        let queuedIDs: Set<String>
        do {
            queuedIDs = Set(try await queue.reports().map(\.reportID))
        } catch {
            return 0
        }

        var recovered = 0
        for record in store.records() {
            if queuedIDs.contains(record.reportID) {
                if store.remove(reportID: record.reportID) { recovered += 1 }
                continue
            }

            guard let envelope = try? CrumbJavaScriptCrashEnvelopeBuilder.build(
                settings: settings,
                record: record,
                runtime: runtime
            ) else {
                continue
            }

            do {
                _ = try await queue.enqueue(envelope: envelope, artifacts: [])
                if store.remove(reportID: record.reportID) { recovered += 1 }
            } catch CrumbReportQueueError.queueFull {
                break
            } catch {
                // Keep the pending handoff when storage is unavailable or corrupt.
            }
        }
        return recovered
    }
}

private struct JavaScriptCrashEnvelopeDTO: Encodable {
    let schemaVersion: String
    let reportID: String
    let trigger: String
    let triggeredAt: Date
    let submittedAt: Date
    let release: ReleaseDTO
    let application: ApplicationDTO?
    let sdk: SDKDTO
    let runtime: RuntimeDTO
    let userInput: UserInputDTO
    let javascriptCrash: JavaScriptCrashDTO
    let diagnostics: DiagnosticsDTO
    let privacy: PrivacyDTO
    let customContext: [String: String]?
    let artifacts: [String]
}

private struct ReleaseDTO: Encodable {
    let platform: String
    let appVersion: String
    let nativeBuild: String
    let jsBundleVersion: String?
    let environment: String
}

private struct ApplicationDTO: Encodable { let name: String }

private struct SDKDTO: Encodable {
    let name: String
    let version: String
    let integration: String
}

private struct RuntimeDTO: Encodable {
    let os: String
    let osVersion: String
    let deviceFamily: String
    let locale: String
    let timezone: String
}

private struct UserInputDTO: Encodable {
    let category: String
    let description: String
}

private struct JavaScriptCrashDTO: Encodable {
    let source: String
    let kind: String
    let errorType: String
    let message: String
    let rawStack: String?
    let deduplicationKey: String
    let nativeTerminationWrapper: Bool
    let breadcrumbs: [BreadcrumbDTO]
}

private struct BreadcrumbDTO: Encodable {
    let timestamp: Date
    let category: String
    let message: String
}

private struct DiagnosticsDTO: Encodable {
    let capturedAt: Date
    let location: String
    let process: ProcessDTO
    let thermalState: String
    let threads: ThreadsDTO
    let gpu: GPUDTO
    let network: NetworkDTO
    let logs: LogsDTO
    let stackTraces: StackTracesDTO
}

private struct ProcessDTO: Encodable {
    let name: String
    let id: Int
}

private struct ThreadsDTO: Encodable {
    let count: Int
    let busiest: [String]
}

private struct GPUDTO: Encodable { let status: String }

private struct NetworkDTO: Encodable {
    let status: String
    let transport: String
    let cellularGeneration: String?
    let isExpensive: Bool
    let isConstrained: Bool
    let healthCheck: String?
}

private struct LogsDTO: Encodable {
    let status: String
    let sources: [String]
    let entries: [String]
    let truncated: Bool
    let droppedEntryCount: Int
    let failures: [String]
}

private struct StackTracesDTO: Encodable {
    let status: String
    let scope: String
    let threads: [String]
    let truncated: Bool
    let unavailableReason: String?
}

private struct PrivacyDTO: Encodable {
    let screenshotCapture: String
    let screenshotMasking: String
    let diagnosticsCapture: String
    let logCapture: String
    let policyStatus: String
    let workspacePolicyVersion: Int?
}

File: packages/android/crumb-core/src/main/kotlin/dev/crumb/core/CrumbJavaScriptCrashEnvelope.kt
package dev.crumb.core

import java.time.Instant

internal object CrumbJavaScriptCrashEnvelopeBuilder {
    fun build(
        settings: CrumbReportSettings,
        record: CrumbJavaScriptCrashRecord,
        runtime: CrumbJavaScriptCrashRuntime,
    ): CrumbSerializedReportEnvelope {
        val crash = record.crash
        val customContext = record.customContext.filter { (key, value) ->
            settings.customContext[key] == value
        }
        val envelope = obj(
            "schema_version" to "1.0",
            "report_id" to record.reportId,
            "trigger" to "javascript_crash",
            "triggered_at" to instant(crash.occurredAtMillis),
            "submitted_at" to instant(crash.occurredAtMillis),
            "release" to obj(
                "platform" to "react_native",
                "app_version" to record.appVersion,
                "native_build" to record.nativeBuild,
                "js_bundle_version" to record.bundleVersion,
                "environment" to record.environment,
            ),
            "application" to record.applicationName?.let { obj("name" to it) },
            "sdk" to obj(
                "name" to "crumb-react-native",
                "version" to CrumbSDKVersion.CURRENT,
                "integration" to "react_native",
            ),
            "runtime" to obj(
                "os" to "android",
                "os_version" to runtime.osVersion,
                "device_family" to runtime.deviceFamily,
                "locale" to runtime.locale,
                "timezone" to runtime.timezone,
            ),
            "user_input" to obj(
                "category" to "javascript_crash",
                "description" to CrumbJavaScriptCrashSanitizer.truncateUtf8(
                    "${crash.errorType}: ${crash.message}",
                    4_000,
                ),
            ),
            "javascript_crash" to obj(
                "source" to crash.source.wireValue,
                "kind" to crash.kind.wireValue,
                "error_type" to crash.errorType,
                "message" to crash.message,
                "raw_stack" to crash.rawStack,
                "deduplication_key" to (crash.fingerprint ?: "js_unknown"),
                "native_termination_wrapper" to crash.nativeTerminationWrapper,
                "breadcrumbs" to crash.breadcrumbs.map { breadcrumb ->
                    obj(
                        "timestamp" to instant(breadcrumb.timestampMillis),
                        "category" to breadcrumb.category,
                        "message" to breadcrumb.message,
                    )
                },
            ),
            "diagnostics" to obj(
                "captured_at" to instant(crash.occurredAtMillis),
                "location" to "javascript_crash_recovery",
                "process" to obj(
                    "name" to runtime.processName,
                    "id" to runtime.processId,
                ),
                "thermal_state" to "unavailable",
                "threads" to obj(
                    "count" to 0,
                    "busiest" to emptyList<Any>(),
                ),
                "gpu" to obj("status" to "unavailable_on_demand"),
                "network" to obj(
                    "status" to "unknown",
                    "transport" to "unknown",
                    "is_expensive" to false,
                    "is_constrained" to false,
                ),
                "logs" to obj(
                    "status" to "unavailable",
                    "sources" to emptyList<String>(),
                    "entries" to emptyList<Any>(),
                    "truncated" to false,
                    "dropped_entry_count" to 0,
                    "failures" to listOf("crash_recovery"),
                ),
                "stack_traces" to obj(
                    "status" to "unavailable",
                    "scope" to "none",
                    "threads" to emptyList<Any>(),
                    "truncated" to false,
                    "unavailable_reason" to "crash_recovery",
                ),
            ),
            "privacy" to obj(
                "screenshot_capture" to "not_applicable",
                "screenshot_masking" to "not_applicable",
                "diagnostics_capture" to "crash_recovery",
                "log_capture" to "not_applicable",
                "policy_status" to settings.policyStatus.wireValue,
                "policy_version" to settings.workspacePolicyVersion,
            ),
            "custom_context" to customContext.takeIf { it.isNotEmpty() },
            "artifacts" to emptyList<Any>(),
        )
        return CrumbSerializedReportEnvelope(
            reportId = record.reportId,
            submittedAtMillis = crash.occurredAtMillis,
            json = JsonValueEncoder.encode(envelope),
        )
    }

    private fun instant(milliseconds: Long): String = Instant.ofEpochMilli(milliseconds).toString()

    private fun obj(vararg values: Pair<String, Any?>): Map<String, Any?> = linkedMapOf<String, Any?>().apply {
        values.forEach { (key, value) -> if (value != null) put(key, value) }
    }
}

internal object CrumbJavaScriptCrashRecovery {
    fun recoverPending(
        store: CrumbJavaScriptCrashStore,
        queue: CrumbReportQueue,
        settings: CrumbReportSettings,
        runtime: CrumbJavaScriptCrashRuntime,
    ): Int {
        val queuedIds = runCatching { queue.reports().map { it.reportId }.toSet() }
            .getOrNull() ?: return 0
        var recovered = 0
        for (record in store.records()) {
            if (record.reportId in queuedIds) {
                if (store.remove(record.reportId)) recovered += 1
                continue
            }
            val envelope = runCatching {
                CrumbJavaScriptCrashEnvelopeBuilder.build(settings, record, runtime)
            }.getOrNull() ?: continue
            try {
                queue.enqueue(envelope, emptyList())
                if (store.remove(record.reportId)) recovered += 1
            } catch (_: CrumbReportQueueException.QueueFull) {
                break
            } catch (_: Exception) {
                // Keep the pending handoff when storage is unavailable or corrupt.
            }
        }
        return recovered
    }
}
